src/file_pipe_log/file_mgr.rs

```diff
@@ -0,0 +1,591 @@
+// Copyright (c) 2017-present, PingCAP, Inc. Licensed under Apache-2.0.
+
+use std::collections::VecDeque;
+use std::io::Write;
+use std::sync::Arc;
+use std::time::Instant;
+
+use crossbeam::utils::CachePadded;
+use log::{error, info};
+use parking_lot::RwLock;
+
+use crate::env::FileSystem;
+use crate::pipe_log::{FileId, FileSeq, LogQueue, Version};
+use crate::{Error, Result};
+
+use super::format::{FileNameExt, LogFileFormat, StaleFileNameExt};
+
+/// Default buffer size for building stale file, unit: byte.
+const LOG_STALE_FILE_BUF_SIZE: usize = 16 * 1024 * 1024;
+/// Default path id for log files. 0 => Main dir
+const LOG_DEFAULT_PATH_ID: PathId = 0;
+
+pub type PathId = usize;
+pub type Paths = [String; 1];
+
+#[derive(Debug)]
+pub struct FileWithFormat<F: FileSystem> {
+    pub handle: Arc<F::Handle>,
+    pub format: LogFileFormat,
+    pub path_id: PathId,
+}
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct FileState {
+    pub first_seq: FileSeq,
+    pub total_len: usize,
+}
+
+pub struct FileList<F: FileSystem> {
+    /// Sequence number of the first file.
+    first_seq: FileSeq,
+    fds: VecDeque<FileWithFormat<F>>,
+}
+
+impl<F: FileSystem> FileList<F> {
+    #[inline]
+    pub fn new(first_seq: FileSeq, fds: VecDeque<FileWithFormat<F>>) -> Self {
+        let first_seq = if fds.is_empty() { 0 } else { first_seq };
+        Self { first_seq, fds }
+    }
+
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.fds.len()
+    }
+
+    #[inline]
+    pub fn span(&self) -> (FileSeq, FileSeq) {
```

Comment:
Why `Option` ? The meaning of `span()` is clear -- "Returns the `seqno` range of current file list." If the file list is empty, we should return (0, 0) to users.

Reply:
If you want to represent null, just use Option. It's the whole purpose of Option. Otherwise you return a meaningful result. (0,0) is not meaningful (because directly using it will break almost all calculations around first_seq, last_seq), it's just a manmade special marker.